### ISSUE TITLE
ORPO doesn't work with multi-gpu (Deepspeed)

### ISSUE THREAD

**Issue by kevinlu1248** [huggingface/trl#3237]
### Reproduction

Here are my configs. Running on 8xH100 on Modal.

```python
trainer = ORPOTrainer(
    **shared_training_args,
    args=ORPOConfig(
        **shared_config_args,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=16,
        eval_accumulation_steps=16,
        warmup_ratio=0.1,
        beta=0.1,
        learning_rate=3e-7,
        weight_decay=0.01,
        lr_scheduler_type="cosine",
        max_prompt_length=max_prompt_length,
        max_completion_length=max_completion_length,
        max_length=max_prompt_length + max_completion_length,
        metric_for_best_model="eval_loss",
    ),
)
```

outputs:

```
{'loss': 5.2878, 'grad_norm': nan, 'learning_rate': 6e-09, 'rewards/chosen': -0.0010004043579101562, 'rewards/rejected': -0.0017032623291015625, 'rewards/accuracies': 0.890625, 'rewards/margins': 0.0007028579711914062, 'logps/rejected': -0.01702117919921875, 'logps/chosen': -0.009983062744140625, 'logits/rejected': 0.68701171875, 'logits/chosen': 0.614990234375, 'nll_loss': 0.5673828125, 'log_odds_ratio': nan, 'log_odds_chosen': nan, 'epoch': 0.06}
```

it's different from the other NaN issues in that it's mostly non-nan except log odds chosen and all it's dependencies. The problem goes away when I run on single GPU.

### System Info

- Platform: Linux-4.4.0-x86_64-with-glibc2.35
- Python version: 3.11.5
- TRL version: 0.17.0.dev0
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA H200
- Transformers version: 4.49.0
- Accelerate version: 1.4.0
- Accelerate config: not found
- Datasets version: 3.3.2
- HF Hub version: 0.29.2
- bitsandbytes version: 0.45.3
- DeepSpeed version: 0.16.4
- Diffusers version: 0.32.2
- Liger-Kernel version: 0.5.4
- LLM-Blender version: not installed
- OpenAI version: 1.65.3
- PEFT version: 0.14.0
- vLLM version: 0.7.3

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by kevinlu1248** [comment]
Found that clamping partially fixes it. Also training with SFT first then ORPO also fixes it, since the initial logprobs wouldn't be out of range.

**Comment by kevinlu1248** [comment]
Had a chat with a couple reasoning models on this and they all proposed the logits should be clamped from around 1e-6 to 25, especially for bf16, can this be provided by default?

**Comment by xiyang-aads-lilly** [comment]
> Found that clamping partially fixes it. Also training with SFT first then ORPO also fixes it, since the initial logprobs wouldn't be out of range.

How did you do the clamping?  I try to clamp the logps using following but I am not sure if this will influence the training performance:
```python
EPS = 1e-6
MAX_LOGIT = torch.log(torch.tensor(1.0 - EPS))

policy_chosen_logps = torch.clamp(policy_chosen_logps, min=-20, max=MAX_LOGIT)
policy_rejected_logps = torch.clamp(
    policy_rejected_logps, min=-20, max=MAX_LOGIT
)
```

**Comment by amit-gupta-** [comment]
@kevinlu1248 Were you able to fix the issue with multi gpu ?

**Comment by hartmans** [comment]
I've had fairly successful deepspeed training on  2 H200s on Runpod.
However, to get that I had to fix a stability problem in the loss function. I actually ran into that loss problem more with single-GPU training than with Deepspeed/FSDP, but I did run into it with Deepspeed too.
Basically any heavy use of ORPO ran into NAN from the loss until I fixed the stability.
See #4763 
I'm curious if applying that patch improves your Deepspeed issues.

**Comment by DaoyuanLi2816** [comment]
This looks like the same catastrophic-cancellation bug in the ORPO loss that #4763 fixed (merged in 2abee9d).

The numbers in the original report match that failure mode exactly: `log_odds_ratio: nan` and `log_odds_chosen: nan` while everything else (`loss`, `rewards/*`, `logps/*`) stays finite. Before #4763, `odds_ratio_loss` computed:

```python
torch.log1p(-torch.exp(policy_chosen_logps))
```

which returns `-inf` whenever `policy_chosen_logps` gets close to `0` (i.e. the model is confident about the chosen response) — `exp(x) -> 1`, so `log1p(-1) = log(0)`. That in turn makes `log_odds` (and everything downstream of it) `nan`. This is more likely to show up under multi-GPU/bf16 training simply because logits differ slightly per-rank and it only takes one confident example to hit the edge case. It also lines up with @kevinlu1248's finding that clamping the logps "partially fixes it," since clamping keeps `policy_chosen_logps` away from that edge.

#4763 replaced this with a numerically stable `log1mexp` (branches between `log1p(-exp(x))` and `log(-expm1(x))` depending on the sign of `x` relative to `-ln 2`) and is bit-identical to the old formula away from the problematic region. That's on current `main` (`trl/experimental/orpo/orpo_trainer.py`).

@hartmans's question upthread — "I'm curious if applying that patch improves your Deepspeed issues" — seems to already be answered by their own report of successful DeepSpeed training after applying it. Given the exact metric match, this looks resolved.

@kevinlu1248 since you opened this — if you're no longer hitting it on current `main` (or with the clamping workaround no longer needed), feel free to close.

**Comment by behroozazarkhalili** [comment]
This looks resolved on current `main`. As @DaoyuanLi2816 noted, the failure here is the ORPO odds-ratio loss suffering catastrophic cancellation, and @hartmans (who hit the same instability on 2x H200 DeepSpeed) fixed it in #4763.

The fix is live in `main`: `odds_ratio_loss` now casts the log-probs to float and computes `log(1 - exp(x))` with a numerically stable `log1mexp` instead of `torch.log1p(-torch.exp(...))`:

```python
policy_chosen_logps = policy_chosen_logps.float()
policy_rejected_logps = policy_rejected_logps.float()
log_odds = (policy_chosen_logps - policy_rejected_logps) - (
    log1mexp(policy_chosen_logps) - log1mexp(policy_rejected_logps)
)
```

The old `log1p(-exp(logp))` cancels catastrophically when a response token has probability close to 1 (`logp -> 0`), producing `-inf`/`nan` in the loss regardless of the number of GPUs. Since the root cause is the loss and not the multi-GPU/DeepSpeed path itself, #4763 fixes it for single- and multi-GPU alike, which matches @hartmans reporting successful DeepSpeed ORPO training after the loss fix.

Note ORPO now lives under `trl.experimental.orpo`. This can likely be closed as fixed by #4763. cc @qgallouedec

**Comment by albertvillanova** [comment]
Hey @kevinlu1248, sorry for the very slow reply here — we're catching up on the backlog.

This should be fixed on current `main`, and the root cause turned out not to be the multi-GPU/DeepSpeed path itself but catastrophic cancellation in the ORPO loss. `odds_ratio_loss` used to compute `log1p(-exp(logp))`, and because ORPO uses per-token *average* log-probs, a confident model puts `logp` very close to 0 — in bf16, `exp(x)` rounds to exactly 1.0 once `|x| <= ~1e-3`, so that term becomes `-inf`, and when chosen and rejected both saturate you get `-inf - (-inf) = nan`. That is exactly your signature: `log_odds_ratio`/`log_odds_chosen` nan while `nll_loss` and `rewards/*` stay finite (the `loss` column looks healthy only because `logging_nan_inf_filter` defaults to `True`, so it silently substitutes the running average). Multi-GPU mostly increased your exposure — 8 ranks x `gradient_accumulation_steps=16` pushes 8x more samples through the loss per step, and one rank's nan poisons the all-reduce.

@hartmans fixed this in #4763 by casting to fp32 and using a numerically stable `log1mexp` (https://github.com/huggingface/trl/blob/c061050d3c4eb62dda345adf6f1a4e892619adec/trl/experimental/orpo/orpo_trainer.py#L88-L92, applied at https://github.com/huggingface/trl/blob/c061050d3c4eb62dda345adf6f1a4e892619adec/trl/experimental/orpo/orpo_trainer.py#L694-L698). Worth noting your clamping workaround was masking a genuinely wrong loss rather than just a nan: at your own reported `logps/chosen = -0.0099` / `logps/rejected = -0.0170`, the old formula returns `log_odds = 0.289` against a float64 reference of `0.544`. Also note ORPO now lives under `trl.experimental.orpo`.

Closing as fixed — but please re-open if you still hit nan on a recent version and we'll dig back in.
